README.md
SCATE: Single-cell ATAC-seq Signal Extraction and Enhancement
====

## Introductions
Single-cell sequencing assay for transposase-accessible chromatin (scATAC-seq) is
the state-of-the-art technology for analyzing genome-wide regulatory landscape in
single cells. Due to data sparsity and discreteness, analyzing scATAC-seq data is
challenging. Existing computational methods cannot accurately reconstruct
activities of individual cis-regulatory elements (CREs) in individual cells. We
present a new statistical framework, SCATE, that adaptively integrates
information from co-activated CREs, similar cells, and publicly available regulome
data to substantially increase the accuracy for estimating individual CRE
activities in single cell and rare cell subpopulations. 

## SCATE Installation

SCATE software can be installed via Github.
Users should have R installed on their computer before installing SCATE. R version needs to be at least 3.5.x or higher. R can be downloaded here: http://www.r-project.org/.

For Windows users, Rtools is also required to be installed. Rtools can be downloaded here: (https://cloud.r-project.org/bin/windows/Rtools/). For R version 3.5.x, Rtools35.exe is recommended. Use default settings to perform the installation.

For mac users, if there is any problem with installation problem, please try download and install clang-8.0.0.pkg from the following URL: https://cloud.r-project.org/bin/macosx/tools/clang-8.0.0.pkg

To install the latest version of SCATE package via Github, run following commands in R:
```{r }
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("GenomicAlignments","preprocessCore"))
if (!require("devtools"))
  install.packages("devtools")
devtools::install_github("zji90/SCATE")
```
<!---
In rare case where devtools cannot be installed, SCATE software package can also be downloaded from this link:

http://jilab.biostat.jhsph.edu/projects/scate/SCATE_1.0.tar.gz

After downloading the package, open R, type in the following commands:
```{r }
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GenomicAlignments")
```
After the installation process, go to 'Packages & Data' menu, go to 'Package Installer', choose 'Local Source Package' from the pull-down menu the top, click 'Install' button on the bottom, choose the path to the software package 'SCATE_1.0.tar.gz' just downloaded, and click 'open'. The installation process will be completed soon. 
--->

If there is any problem with the installation process, please make sure you have R version at least 3.5.x and you have installed Rtools (Windows users) or clang (mac users). If the problem still occurs, please contact the author (see below)

## User Manual
Check the following page for PDF version of the user manual:
https://github.com/zji90/SCATE/raw/master/inst/doc/SCATE.pdf

Check below link for the R code of the user manual.

https://github.com/zji90/SCATE/blob/master/inst/doc/SCATE.R

## Citation 

Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji, [*Single-cell ATAC-seq Signal Extraction and Enhancement with SCATE*](https://doi.org/10.1186/s13059-020-02075-3), Genome Biol 21, 161 (2020).

## Contact
Author: Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji 

Report bugs and provide suggestions by sending email to:

Maintainer: Zhicheng Ji (zji4@jhu.edu), Wenpin Hou (whou10@jhu.edu)

Or open a new issue on this Github page


inst/NEWS
Changes in version 1.0.0
------------------------

  - submission to Bioconductor


R/peakcall.R
#' Peak calling
#'
#' Peak calling function 
#'
#' This function performs peak calling for signal generated by SCATE
#' @param res Result matrix returned by SCATE
#' @param flank Numeric variable of the number of flanking bins for each bin. For each bin, an averaged signal of itself and the flanking bins will be calculated and compared to a background distribution.
#' @param fdrcut Numeric variable of FDR cutoff. Bins passing the FDR cutoff will be peaks.
#' @return A list with length equal to the number of clusters. Each element is a data frame with five columns: chromosome name, starting location, ending location, FDR and signal. The data frame is ordered by FDR then by signal.
#' @export
#' @import GenomicAlignments GenomicRanges
#' @author Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji* <whou10@@jhu.edu>
#' @examples
#' gr <- GRanges(seqnames="chr1",IRanges(start=seq_len(100)+1e6,end=seq_len(100)+1e6))
#' scateout <- SCATE(gr,clunum=156,genome='mm10')[seq_len(1000000),,drop=FALSE]
#' peakcall(scateout)

peakcall <- function(res,flank=1,fdrcut=1e-5) {
   gr <- strsplit(row.names(res),'_')
   gr <- do.call(rbind,gr)
   gr <- GRanges(seqnames=gr[,1],IRanges(start=as.numeric(gr[,2]),end=as.numeric(gr[,3])))
   flankgr <- gr
   start(flankgr) <- start(flankgr) - 200 * flank
   end(flankgr) <- end(flankgr) + 200 * flank
   flankgrover <- as.matrix(findOverlaps(gr,flankgr))
   
   peakres <- lapply(seq(1, ncol(res)),function(i) {
      feature <- res[,i]
      sumv <- rowsum(feature[flankgrover[,1]],flankgrover[,2])[,1]
      sumv <- sumv/rowsum(rep(1,nrow(flankgrover)),flankgrover[,2])[,1]
      
      sampfeature <- sample(feature)
      backv <- sapply(seq(0,100000),function(i) mean(sampfeature[(1+i*(2*flank+1)):((i+1)*(2*flank+1))]))
      xseq <- seq(floor(min(sumv)),ceiling(max(sumv)),0.1)
      fdr <- sapply(xseq,function(cut) {
         mean(backv >= cut)/mean(sumv >= cut)
      })
      fdr <- cummin(fdr)
      fdr <- approx(xseq[!is.na(fdr)],fdr[!is.na(fdr)],sumv,rule=2)$y
      sig <- which(fdr < fdrcut)
      centergr <- gr[sig]
      
      boundgr <- gr[unique(flankgrover[flankgrover[,2] %in% sig,1])]
      boundgr <- reduce(boundgr)
      over <- as.matrix(findOverlaps(centergr,boundgr))
      fdr <- tapply(fdr[sig][over[,1]],list(over[,2]),min)
      signal <- tapply(sumv[sig][over[,1]],list(over[,2]),max)
      dat <- data.frame(chr=as.character(seqnames(boundgr)),start=start(boundgr),end=end(boundgr),FDR=fdr,Signal=signal)
      dat[order(dat$FDR,-dat$Signal),]
   })
   names(peakres) <- colnames(res)
   peakres
}



R/cellcluster.R
#' Cell clustering
#'
#' Perform Cell Clustering
#'
#' This function generates averaged signals for CRE clusters and cluster cells.
#' @param satac If type='reads', satac should be a list of GRanges object of scATAC-seq reads. Each element corresponds to one single cell. The GRanges should be the middle point of the reads with length of 1 base pair. Use 'satacprocess' to preprocess raw reads. If type='peaks', satac should be a list of data frames of scATAC-seq peaks. For each data frame, first column is chromsome name, second column is start site, third column is end site, and fourth column is the number of reads of the peak.
#' @param type Character variable of either 'reads' or 'peaks'.
#' @param peakOverlapMethod Character variable of either 'full' or 'middle'. Only effective when type = 'peaks'. If peakOverlapMethod='full', then the full range of the peak will be used to find overlap with bins, and all bins overlapping with this peak will be assigned the read counts of this peak. If peakOverlapMethod='middle', only the middle base pair of the peak will be used to find overlap with bins.
#' @param genome Character variable of either "hg19" or "mm10".
#' @param clunum Numeric variable giving the number of clusters. If NULL the cluster number will be determined automatically
#' @param perplexity Numeric variable specifying perplexity for tSNE. Reduce perplexity when sample size is small.
#' @param filtervar If TRUE, filter out features with low variability.
#' @param datapath Character variable of the path to the customized database (eg myfolder/database.rds). The database can be made using 'makedatabase' function. If not null, 'genome' is ignored.
#' @return A list of three components: tsne results, clustering results and aggregated signal for CRE cluster.
#' @export
#' @import GenomicAlignments GenomicRanges mclust splines Rtsne preprocessCore
#' @author Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji* <whou10@@jhu.edu>
#' @examples
#' celldata <- lapply(seq_len(50),function(i) {
#'   pos <- sample(seq_len(1e9),50000)
#'   GRanges(seqnames=sample(paste0("chr",seq_len(20)),50000,replace=TRUE),IRanges(start=pos,end=pos))
#' })
#' names(celldata) <- paste0('cell',seq_len(50))
#' cellcluster(celldata,type='reads',genome="hg19",filtervar=FALSE,perplexity=1,clunum=3) # reads as input

cellcluster <- function(satac,type='reads',peakOverlapMethod = 'full',genome='hg19',clunum=NULL,perplexity=30,filtervar=TRUE,datapath=NULL) {
    if (!is.null(datapath)) {
            loaddata <- readRDS(datapath)
    } else {
            loaddata <- readRDS(paste0(system.file(package="SCATEData"),"/extdata/",genome,".rds"))
    }
   allclunum <- loaddata$allclunum
   clu <- loaddata$cluster[,allclunum==5000]
   tabclu <- as.vector(table(clu))
   gr <- loaddata$gr
   id <- loaddata$id
   gr <- gr[id]
   clusteraggregate <- sapply(names(satac),function(sid) {
         if (type=='reads') {
               len <- length(satac[[sid]])/1e8
               rowsum(log2(countOverlaps(gr,satac[[sid]],ignore.strand=TRUE)/len+1),clu)/tabclu
         } else if (type=='peaks') {
               len <- sum(satac[[sid]][,4])/1e8
               if (peakOverlapMethod == 'full'){
                     peak <- GRanges(seqnames=satac[[sid]][,1],IRanges(start=satac[[sid]][,2],end=satac[[sid]][,3]))
               } else {
                     tmp <- floor((satac[[sid]][,2]+satac[[sid]][,3])/2)
                     peak <- GRanges(seqnames=satac[[sid]][,1],IRanges(start=tmp,end=tmp))
               }
               o <- as.matrix(findOverlaps(gr,peak))
               count <- rep(0,length(gr))
               count[o[,1]] <- satac[[sid]][o[,2],4]
               rowsum(log2(count/len+1),clu)/tabclu
         } else {
               stop('Wrong type')
         }
   })
   
   row.names(clusteraggregate) <- seq(1,nrow(clusteraggregate))
   clusteraggregate <- clusteraggregate[rowMeans(clusteraggregate > 0) > 0.1,]
   dn <- dimnames(clusteraggregate)
   clusteraggregate <- normalize.quantiles(clusteraggregate)
   dimnames(clusteraggregate) <- dn
   if (filtervar) {
         rsd <- apply(clusteraggregate,1,sd)
         rm <- rowMeans(clusteraggregate)
         clusteraggregate <- clusteraggregate[rsd > fitted(lm(rsd~bs(rm))),,drop=FALSE]
   }
   prres <- prcomp(t(clusteraggregate),scale. = TRUE)$x
   
   tsne <- Rtsne(prres[,seq(1,min(50,ncol(prres)))],pca=FALSE,perplexity=perplexity)$Y
   row.names(tsne) <- row.names(prres)
   
   if (is.null(clunum)) {
         cluster <- Mclust(tsne,G=seq(1,20),prior = priorControl(),verbose=FALSE)
   } else {
         cluster <- Mclust(tsne,G=clunum,prior = priorControl(),verbose=FALSE)
   }
   cluster <- apply(cluster$z,1,which.max)
   names(cluster) <- row.names(tsne)
   list(tsne=tsne,cluster=cluster,clusteraggregate=clusteraggregate)
}



R/makedatabase.R
#' Make customized database
#'
#' Make customized database with new bulk DNase-seq data
#'
#' This function makes a new customized database if users have new bulk DNase-seq data and such information can be contribued to the model building of SCATE.
#' @param datapath path to the data package folder (e.g. myfolder/hg19/). User must first download the data package to use this function. The data package for hg19 and mm10 can be downloaded from http://jilab.biostat.jhsph.edu/projects/scate/hg19.zip or http://jilab.biostat.jhsph.edu/projects/scate/mm10.zip. The compressed file should be unzipped. If users do not want to use existing data compendium (e.g. to build a database in a new species), datapath should be set NULL, and 'genome' will be ignored.
#' @param savepath path to save the generated database. e.g. myfolder/database.rds.
#' @param blacklist GRanges object that identifies blacklisted regions to be filtered out.
#' @param bamfile location of bulk DNase-seq bamfiles.
#' @param cre dataframe of new CRE sites to be added to the database. First column: chromosome name. Second column: start position. Third column: end position.
#' @param genome Character variable of either "hg19" or "mm10". Default is 'hg19'. Ignored when datapath is NULL.
#' @param genomerange Data frame with two columns. First column is the chromosome and second column is the length of the genome. Only useful when datapath is NULL. Example is https://genome.ucsc.edu/goldenpath/help/hg19.chrom.sizes
#' @return a new customized database if users have new bulk DNase-seq data and such information can be contribued to the model building of SCATE.
#' @export
#' @import GenomicAlignments GenomicRanges
#' @author Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji* <whou10@@jhu.edu>

makedatabase <- function(datapath,savepath,blacklist=NULL,bamfile=NULL,cre=NULL,genome='hg19',genomerange=NULL) {
   if (is.null(datapath)) {
            chrlen <- genomerange
            allres <- NULL
            #sort chromosome
            for (i in seq(1, nrow(chrlen))) {
                  #keep all chromosome
                  start <- 200 * (0:floor((chrlen[i,2])/200))
                  end <- start + 199
                  allres <- rbind(allres,data.frame(chr=chrlen[i,1],start=start,end=end))
            }
            gr <- GRanges(seqnames=allres$chr,IRanges(start=allres$start,end=allres$end))
            over <- as.matrix(findOverlaps(gr,blacklist))[,1]
            gr <- gr[setdiff(seq(1, length(gr)),over),]

            if (!is.null(cre)) {
                  cre <- GRanges(seqnames=cre[,1],IRanges(start=cre[,2],end=cre[,3]))
                  creid <- as.matrix(findOverlaps(gr,cre))[,1]	
            } else {
                  creid <- NULL
            }
            
            len <- rep(0,length(bamfile))
            names(len) <- bamfile
            count <- matrix(0,nrow=length(gr),ncol=length(bamfile),dimnames = list(NULL,bamfile))
            for (f in bamfile) {
                  tmp <- readGAlignmentPairs(f)
                  if (length(tmp) == 0) tmp <- readGAlignments(f)
                  tmp <- GRanges(tmp)
                  mid <- round((start(tmp) + end(tmp))/2)
                  tmp <- GRanges(seqnames=as.character(seqnames(tmp)),IRanges(start=mid,end=mid))
                  count[,f] <- countOverlaps(gr,tmp)
                  len[f] <- length(tmp)
            }
            rm('tmp')
            len <- len/1e8
            norm <- log2(t(t(count)/len)+1)
            
            
            start <- end <- rep(0,length(gr))
            sn <- as.character(seqnames(gr))
            for (s in seqlevels(gr)) {
                  id <- which(sn==s)
                  start[id] <- pmax(min(id),id - 250)
                  end[id] <- pmin(max(id),id + 250)
            }
            
            pos <- cbind(start,end)
            
            newid <- lapply(seq(1, ncol(count)),function(i) {  
                  id <- which(count[,i] >= 10 & norm[,i] >= 5)
                  backnorm <- sapply(id,function(j) mean(norm[pos[j,1]:pos[j,2],i]))
                  id[which(norm[id,i]/backnorm >= 5)]	
            })
            
            id <- unique(c(unlist(newid),creid))
      } else {
            packdata <- readRDS(paste0(system.file(package="SCATEData"),"/extdata/",genome,".rds"))
            gr <- packdata$gr
            
            if (!is.null(cre)) {
                  cre <- GRanges(seqnames=cre[,1],IRanges(start=cre[,2],end=cre[,3]))
                  creid <- as.matrix(findOverlaps(gr,cre))[,1]	
            } else {
                  creid <- NULL
            }
            
            len <- rep(0,length(bamfile))
            names(len) <- bamfile
            count <- matrix(0,nrow=length(gr),ncol=length(bamfile),dimnames = list(NULL,bamfile))
            for (f in bamfile) {
                  tmp <- readGAlignmentPairs(f)
                  if (length(tmp) == 0) tmp <- readGAlignments(f)
                  tmp <- GRanges(tmp)
                  mid <- round((start(tmp) + end(tmp))/2)
                  tmp <- GRanges(seqnames=as.character(seqnames(tmp)),IRanges(start=mid,end=mid))
                  count[,f] <- countOverlaps(gr,tmp)
                  len[f] <- length(tmp)
            }
            rm('tmp')
            len <- len/1e8
            norm <- log2(t(t(count)/len)+1)
            
            pos <- readRDS(paste0(datapath,"/backgrpos.rds"))
            newid <- lapply(seq(1, ncol(count)),function(i) {  
                  id <- which(count[,i] >= 10 & norm[,i] >= 5)
                  backnorm <- sapply(id,function(j) mean(norm[pos[j,1]:pos[j,2],i]))
                  id[which(norm[id,i]/backnorm >= 5)]	
            })
            
            id <- unique(c(packdata$id,unlist(newid),creid))
            
            dcount <- readRDS(paste0(datapath,'/countmat.rds'))
            dlen <- readRDS(paste0(datapath,'/len.rds'))/1e8
            dnorm <- log2(t(t(dcount)/dlen)+1)
            count <- cbind(count,dcount)
            len <- c(len,dlen)
            norm <- cbind(norm,dnorm)
            rm('dcount','dlen','dnorm')      
      }
      
      zid <- which(rowSums(norm) == 0)
      id <- setdiff(id,zid)
      excid <- setdiff(seq(1,length(gr)),c(id,zid))
      
      orisum <- sum1 <- sum2 <- sum3 <- rep(0,length(gr))
      for (f in colnames(count)) {
            orisum <- orisum + count[,f]
            sum1 <- sum1 + count[,f]/len[f]
            sum2 <- sum2 + count[,f]^2/len[f]^2
            sum3 <- sum3 + count[,f]/len[f]^2
      }
      mean1 <- sum1/ncol(count)
      mean2 <- sum2/ncol(count)
      mean3 <- sum3/ncol(count)
      alls <- sqrt(log(pmax(1,(mean2-mean3)/mean1^2)))
      allm <- log(mean1)-alls^2/2
      allm <- round(allm,digits=4)
      alls <- round(alls,digits=4)
      s <- alls[id]
      m <- allm[id]
      excs <- alls[excid]
      excm <- allm[excid]
      
      norm <- norm[id,]
      rm('alls','allm','count','sum1','sum2','sum3','orisum','mean1','mean2','mean3')
      scalematrix <- function(data) {
            cm <- rowMeans(data)
            csd <- apply(data, 1, sd)
            (data - cm) / csd
      }
      
      data <- scalematrix(norm)
      data <- round(data,digits=4)
      
      oriclu <- kmeans(data,5000)$cluster
      
      splitclu <- lapply(seq(1,max(oriclu)),function(cid) {
            distmat <- dist(norm[oriclu==cid,])
            hclust(distmat)
      })
      
      clulist <- list()
      clulist[['5000']] <- oriclu
      
      options(scipen=999)
      spclu <- split(seq(1,length(oriclu)),oriclu)
      names(spclu) <- NULL
      
      for (split in c(2, 4, 8, 16, 32, 64)) {
            cluster <- oriclu
            curclu <- 1
            for (i in seq(1, max(oriclu))) {
                  if (sum(oriclu==i) == 1) {
                        cluster[oriclu==i] <- curclu
                        curclu <- curclu + 1
                  } else {
                        tmpclunum <- min(sum(oriclu==i),split)
                        tmpcluster <- cutree(splitclu[[i]],tmpclunum)
                        for (j in seq(1, tmpclunum)) {
                              cluster[spclu[[i]][tmpcluster==j]] <- curclu
                              curclu <- curclu + 1
                        }
                  }
            }
            clulist[[as.character(max(cluster))]] <- cluster
      }
      
      options(scipen=999)
      clucenter <- matrix(0,nrow=max(oriclu),ncol=ncol(norm))
      for (i in seq(1,max(oriclu))) {
            clucenter[i,] <- colMeans(norm[which(oriclu==i),,drop=FALSE])
      }
      
      hclu <- hclust(dist(scalematrix(clucenter)))
      
      for (clufac in c(2, 4, 8, 16, 32)) {
            clunum <- round(max(oriclu) / clufac)
            clu <- cutree(hclu,k = clunum)
            cluster <- rep(0,length(oriclu))
            for (i in seq(1,max(oriclu))) {
                  cluster[which(oriclu==i)] <- clu[i]
            }
            clulist[[as.character(clunum)]] <- cluster
      }
      
      cluster <- do.call(cbind,clulist)
      allclunum <- as.numeric(colnames(cluster))
      data <- list(ms=list(m=m,s=s),id=id,gr=gr,cluster=cluster,excid=excid,excms=list(m=excm,s=excs),allclunum=allclunum)
      saveRDS(data,file=savepath)
}



R/SCATE.R
#' Perform SCATE
#'
#' Single-cell ATAC-seq signal Extration and Enhancement
#'
#' This function takes as input the scATAC-seq reads and generates enhanced signals. Users can either perform SCATE on clusters of cells or a single group of cells.
#' @param satac If type='reads', satac should be a GRanges object or list of GRanges object of scATAC-seq reads. Each element corresponds to one single cell. The GRanges should be the middle point of the reads with length of 1 base pair. Use 'satacprocess' to preprocess raw reads. If type='peaks', satac should be a data frame or list of data frames of scATAC-seq peaks. For each data frame, first column is chromsome name, second column is start site, third column is end site, and fourth column is the number of reads of the peak.
#' @param type Character variable of either 'reads' or 'peaks'.
#' @param peakOverlapMethod Character variable of either 'full' or 'middle'. Only effective when type = 'peaks'. If peakOverlapMethod='full', then the full range of the peak will be used to find overlap with bins, and all bins overlapping with this peak will be assigned the read counts of this peak. If peakOverlapMethod='middle', only the middle base pair of the peak will be used to find overlap with bins.
#' @param genome Character variable of either "hg19" or "mm10". Default is 'hg19'.
#' @param cluster Numeric vector specifying the cluster of cells. Needs to be named and include all cells in satac. If NULL, SCATE will be run on all cells in satac.
#' @param clusterid Numeric number specifying the single cluster to run SCATE. If NULL SCATE will be run on all clusters. Ignored if cluster is NULL. The cluster id must be included in variable 'cluster'.
#' @param clunum Numeric value specifying number of CRE clusters. If NULL, SCATE automatically chooses number of CRE clusters.
#' @param datapath Character variable of the path to the customized database (eg myfolder/database.rds). The database can be made using 'makedatabase' function. If not null, 'genome' is ignored.
#' @param ncores Numeric variable of number of cores to use. If NULL, the maximum number of cores is used.
#' @param verbose Either TRUE or FALSE. If TRUE, progress will be displayed.
#' @return A numeric vector or matrix of values generated by SCATE, depending on the number of clusters. The length of the vector or the number of rows of the matrix is the same as the number of bins in the genome. If a matrix, the column names indicate the cluster id.
#' @export
#' @import GenomicAlignments GenomicRanges parallel splines2 xgboost
#' @author Zhicheng Ji, Weiqiang Zhou, Wenpin Hou, Hongkai Ji* <whou10@@jhu.edu>
#' @examples
#' #Reads as input, setting CRE cluster number as 156 to increase speed. Users need to set it to be NULL in real applications.
#' gr <- GRanges(seqnames="chr1",IRanges(start=seq_len(100)+1e6,end=seq_len(100)+1e8))
#' SCATE(gr,clunum=156,type='reads',genome="mm10") 
#' \dontrun{
#' peak <- data.frame(seqnames="chr1",start=seq_len(100)+1e6,end=seq_len(100)+1e8,count=1)
#' #Peak as input, peakOverlapMethod=full
#' SCATE(satac=peak,clunum=156,type='peaks',genome="mm10") 
#' #Peak as input, peakOverlapMethod=middle
#' SCATE(satac=peak,clunum=156,type='peaks',peakOverlapMethod='middle',genome="mm10") 
#' }

SCATE <- function(satac,type='reads',peakOverlapMethod = 'full',genome='hg19',cluster=NULL,clusterid=NULL,clunum=NULL,datapath=NULL,verbose=TRUE,ncores = 1) {
      if (Sys.info()[['sysname']]=='Windows') {
            message('Parallel is disabled for Windows. Running with one core')
            ncores <- 1
      }
      if (!is.null(datapath)) {
            loaddata <- readRDS(datapath)
      } else {
            loaddata <- readRDS(paste0(system.file(package="SCATEData"),"/extdata/",genome,".rds"))
      }
      gr <- loaddata$gr
      SCATEsingle <- function(satac,genome='hg19',datapath=NULL) {
            options(scipen=999)
            if (verbose) {
                  message('Preparing data')      
            }
            allclunum <- loaddata$allclunum
            ms <- loaddata$ms
            m <- ms$m
            s <- ms$s
            id <- loaddata$id
            gr <- loaddata$gr
            if (type=='reads') {
                  satac <- unlist(GRangesList(satac))
                  count <- countOverlaps(gr[id],satac)
            } else if (type=='peaks') {
                  if (!is.data.frame(satac)) {
                        for (i in seq_len(length(satac)))
                              colnames(satac[[i]]) <- c('chr','start','end','count')
                        satac <- do.call(rbind,satac)
                  }
                  if (peakOverlapMethod == 'full'){
                        peak <- GRanges(seqnames=satac[,1],IRanges(start=satac[,2],end=satac[,3]))
                  } else {
                        tmp <- (satac[,2]+satac[,3])/2
                        peak <- GRanges(seqnames=satac[,1],IRanges(start=tmp,end=tmp))
                  }
                  o <- as.matrix(findOverlaps(gr[id],peak))
                  count <- rep(0,length(id))
                  count[o[,1]] <- satac[o[,2],4]
            } else {
                  stop('Wrong type')
            }
            
            cutm <- as.numeric(cut(m,quantile(m,c(seq(0,10))/10),include.lowest = TRUE))
            hkid <- unlist(lapply(unique(cutm),function(i) {
                  tmpid <- which(cutm==i)
                  tmpid[which(s[tmpid] <= sort(s[tmpid])[1000])]
            }))
            
            mfit <- function(x,y) {
                  tryid <- seq(0,5)
                  BIC <- rep(0,length(tryid))
                  fit <- allkn <- list()
                  for (knotid in seq(1,length(tryid))) {
                        knotnum <- tryid[knotid]
                        if (knotnum==0) {
                              knots <- NULL
                        } else {
                              knots <- quantile(x,seq(0,1,length.out=knotnum+2)[-c(1,knotnum+2)])
                        }
                        base <- cbind(1,iSpline(x,knots=knots,Boundary.knots=range(x)))
                        allkn[[knotid]] <- knots
                        optimfunc <- function(k) {
                              logmu <- (base %*% matrix(k,ncol=1))[,1]
                              -sum(y*logmu-exp(logmu))
                        }
                        fit[[knotid]] <- optim(rep(1,ncol(base)),optimfunc,lower=c(-Inf,rep(0,ncol(base)-1)),method="L-BFGS-B")
                        logmu <- (base %*% matrix(fit[[knotid]]$par,ncol=1))[,1]
                        BIC[knotid] <- log(length(y))*length(fit[[knotid]]$par)-2*sum(y*logmu-exp(logmu))
                  }
                  par <- fit[[which.min(BIC)]]$par
                  knots <- allkn[[which.min(BIC)]]
                  xseq <- seq(min(x),max(x),0.005)
                  base <- cbind(1,iSpline(xseq,knots=knots,Boundary.knots=range(x)))
                  logmu <- (base %*% matrix(par,ncol=1))[,1]
                  
                  slopeleft <- (logmu[length(xseq)*0.1] - logmu[1])/(xseq[length(xseq)*0.1] - xseq[1])
                  sloperight <- (logmu[length(xseq)] - logmu[length(xseq)*0.9])/(xseq[length(xseq)] - xseq[length(xseq)*0.9])
                  interleft <- logmu[1] - slopeleft*xseq[1]
                  interright <- logmu[length(xseq)] - sloperight*xseq[length(xseq)]
                  list(xseq = xseq, logmu = logmu, parleft = c(interleft,slopeleft), parright = c(interright,sloperight))
            }
            
            mfitres <- mfit(m[hkid],count[hkid])
            
            logmufunc <- function(x) {
                  id <- round((x-min(mfitres$xseq))*200) + 1
                  res <- rep(0,length(id))
                  sid <- which(id >= 1 & id <= length(mfitres$xseq))
                  res[sid] <- mfitres$logmu[id[sid]]
                  sid <- which(id < 1)
                  res[sid] <- mfitres$parleft[1] + mfitres$parleft[2] * x[sid]
                  sid <- which(id > length(mfitres$logmu))
                  res[sid] <- mfitres$parright[1] + mfitres$parright[2] * x[sid]
                  res
            }
            
            deltafitfunc <- function(count,m,s) {
                  if (length(count) == 1) {
                        if (count == 0) {
                              optimfunc <- function(delta) {
                                    exp(logmufunc(m+s*delta)) + delta^2/2
                              }
                        } else {
                              optimfunc <- function(delta) {
                                    logmu <- logmufunc(m+s*delta)
                                    -(count*logmu-exp(logmu)) + delta^2/2
                              }
                        }
                  } else {
                        if (sum(count) == 0) {
                              optimfunc <- function(delta) {
                                    sum(exp(logmufunc(m+s*delta))) + delta^2/2
                              }
                        } else {
                              optimfunc <- function(delta) {
                                    logmu <- logmufunc(m+s*delta)
                                    -sum(count*logmu-exp(logmu)) + delta^2/2
                              }
                        }
                  }
                  optimise(optimfunc,c(-100,100))$minimum 
            }
            
            if (verbose) {
                  message('Fitting model')      
            }
            if (is.null(clunum)) {
                  cluster <- loaddata$cluster[,ncol(loaddata$cluster)]
                  spclu <- split(seq(1,length(cluster)),cluster)
                  tabclu <- sapply(spclu,length)
                  targetcluid <- which(tabclu >= 10)
                  targetid <- unlist(spclu[targetcluid])
                  if (length(targetid) > 10000) {
                        targetid <- sample(targetid,10000) 
                  }
                  loglike <- sapply(allclunum,function(clunum) {
                        cluster <- loaddata$cluster[,which(allclunum==clunum)]
                        spclu <- split(seq(1,length(cluster)),cluster)
                        names(spclu) <- NULL
                        if (ncores == 1){
                           loglike <- unlist(lapply(targetid,function(testid) {
                                 trainid <- setdiff(spclu[[cluster[testid]]],testid)
                                 delta <- deltafitfunc(count[trainid],m[trainid],s[trainid])
                                 logmu <- logmufunc(m[testid]+s[testid]*delta)
                                 count[testid]*logmu-exp(logmu)
                           }))
                        } else {
                           loglike <- unlist(mclapply(targetid,function(testid) {
                                 trainid <- setdiff(spclu[[cluster[testid]]],testid)
                                 delta <- deltafitfunc(count[trainid],m[trainid],s[trainid])
                                 logmu <- logmufunc(m[testid]+s[testid]*delta)
                                 count[testid]*logmu-exp(logmu)
                           }, mc.cores = ncores))
                        }
                           
                  })
                  logclu <- log2(c(allclunum,length(id)))
                  medll <- apply(loglike,2,median)
                  loessmod <- loess(medll~logclu,data.frame(medll=medll,logclu=logclu[-length(logclu)]),surface="direct")
                  clunum <- c(allclunum, length(id))[which.max(predict(loessmod,logclu))]
            }
            if (clunum == length(id)) {
                  if (ncores == 1){
                     crefeature <- unlist(lapply(seq(1,length(id)),function(i) {
                           m[i]+s[i]*deltafitfunc(count[i],m[i],s[i])
                     }))
                  } else {
                     crefeature <- unlist(mclapply(seq(1,length(id)),function(i) {
                           m[i]+s[i]*deltafitfunc(count[i],m[i],s[i])
                     }, mc.cores = ncores))
                  }
                     
            } else {
                  cluster <- loaddata$cluster[,which(allclunum==clunum)]
                  spclu <- split(seq(1,length(cluster)),cluster)
                  if (ncores == 1){
                     delta <- unlist(lapply(seq(1,clunum),function(cluid) {
                           binid <- spclu[[cluid]]
                           deltafitfunc(count[binid],m[binid],s[binid])
                     }))
                  } else {
                     delta <- unlist(mclapply(seq(1,clunum),function(cluid) {
                           binid <- spclu[[cluid]]
                           deltafitfunc(count[binid],m[binid],s[binid])
                     }, mc.cores = ncores))
                  }
                     
                  crefeature <- m+s*delta[cluster]
            }
            excid <- loaddata$excid
            excms <- loaddata$excms
            excm <- excms$m
            excs <- excms$s
            if (type=='reads') {
                  exccount <- countOverlaps(gr[excid],satac)
            } else if (type=='peaks') {
                  o <- as.matrix(findOverlaps(gr[excid],peak))
                  exccount <- rep(0,length(excid))
                  exccount[o[,1]] <- satac[o[,2],4]
            }
            logcount <- log2(count + 1)
            logexccount <- log2(exccount + 1)
            
            mod <- xgboost(cbind(m=m,logcount=logcount,s=s),crefeature,nrounds=50,verbose=FALSE)
            predfeature <- predict(mod,cbind(m=excm,logcount=logexccount,s=excs))
            fullfeature <- rep(0,length(gr))
            fullfeature[id] <- crefeature
            fullfeature[excid] <- predfeature
            pmax(0,fullfeature)
      }
      if (is.null(cluster)) {
            res <- matrix(SCATEsingle(satac,genome=genome,datapath=datapath),ncol=1)
            colnames(res) <- 'combine'
      } else {
            cluster <- cluster[names(satac)]
            if (is.null(clusterid)) {
                  target <- sort(unique(cluster))
            } else {
                  target <- sort(intersect(clusterid,cluster))
            }
            res <- sapply(target,function(i) {
                  SCATEsingle(satac[cluster==i],genome=genome,datapath=datapath)
            })     
            colnames(res) <- target
      }
      if (verbose) {
            message('Generating results')      
      }
      row.names(res) <- sprintf('%s_%s_%s',as.character(seqnames(gr)),start(gr),end(gr))
      res
}
